app/src/main/java/io/github/pietrocaselani/moviestraker/ui/searchmovies/SearchMoviesViewModel.kt

```diff
@@ -0,0 +1,192 @@
+import io.github.pietrocaselani.moviestraker.tmdb.entities.ImageConfigurationResponse
+import io.github.pietrocaselani.moviestraker.ui.common.MovieListViewModel
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.CompositeDisposable
+
+/**
+ * Created by pc on 25/06/17.
+ */
+class SearchMoviesViewModel(private val interactor: SearchMoviesInteractorInput) : BaseObservable() {
+
+	//region Data Binding Properties
+	val movies = ObservableField<MutableList<MovieListViewModel>>(mutableListOf())
+	val message = ObservableField<String>("")
+	val resourceMessage = ObservableInt(0)
+	val moviesVisibility = ObservableField<Boolean>(true)
+	val messageVisibility = ObservableField<Boolean>(true)
+	val selectedMovie = ObservableField<MovieEntity>()
+	//endregion
+
+	//region Properties
+	private val disposables = CompositeDisposable()
+
+	private var imageConfiguration: ImageConfigurationResponse? = null
+	private val movieEntities = mutableListOf<MovieEntity>()
+	private var genres = listOf<GenreResponse>()
+	private val loadedPages = mutableListOf<Int>()
+	private var currentPage = 1
+	private var totalPages = Int.MAX_VALUE
+	private var loading = false
+	private var query = ""
+	//endregion
+
+	//region Lifecycle
+	fun onStart() {
+		if (movieEntities.size == 0) {
+			showMessage(R.string.type_movie_name)
+		}
+
+		fetchGenres()
+	}
+
+	fun onStop() {
+		disposables.clear()
+	}
+	//endregion
+
+	//region Data Binding Methods
+	fun search(query: String) {
+		this.query = query
+		this.currentPage = 1
+		loadedPages.clear()
+		loadMovies()
+	}
+	//endregion
+
+	//region Public
+	fun requestMoreMovies() {
+		loading = true
+		currentPage++
+		loadMovies()
+	}
+
+	fun selectMovieAtIndex(index: Int) {
+		val movie = movieEntities[index]
+
+		if (selectedMovie.get() != null && movie.id == selectedMovie.get().id) {
+			selectedMovie.notifyChange()
+		} else {
+			val movieToShow = mapMovieToDetails(movie, imageConfiguration)
+
+			selectedMovie.set(movieToShow)
+		}
+	}
+	//endregion
+
+	//region Private
+	private fun loadMovies() {
+		if (genres.isEmpty()) {
+			fetchGenres()
+			return
+		} else if (imageConfiguration == null) {
+			fetchConfigurations()
+			return
+		}
+
+		if (currentPage >= totalPages || loadedPages.contains(currentPage)) {
+			return
+		}
+
+		loading = true
+
+		val disposable = interactor.searchMovies(query, currentPage)
+				.doOnNext {
+					loadedPages.add(it.page)
+					totalPages = it.totalPages
+				}
+				.map {
+					mapToMovieEntity(it.results, genres, imageConfiguration)
+				}
+				.doOnNext {
+					movieEntities.addAll(it)
+				}
+				.map {
+					mapToMovieListViewModel(it)
+				}
+				.observeOn(AndroidSchedulers.mainThread())
+				.subscribe({
+					showMovies(it)
+				}, {
+					loading = false
+					showMessage(it.localizedMessage)
+				}, {
+					loading = false
+				})
+
+		disposables.add(disposable)
+	}
+
+	private fun fetchGenres() {
+		if (genres.isEmpty()) {
```

Comment:
I've seen this same logic here and there: if genres list is empty then fetch genres, else if configuration is null fetch configurations. 

Is there a way to not repeat it?